Implement new transaction list design

brad-decker opened the pull request:
<details>
<summary>Main View ScreenShot</summary>
<img src="https://user-images.githubusercontent.com/4448075/81867798-e7ec0a00-9536-11ea-8d65-4d33b4f1a21a.png" alt="image" style="max-width:100%;">
</details>

<details>
<summary>Demo Videos</summary>
<a target="__blank" href="https://drive.google.com/file/d/1oyIdWC3kNPARVCATP75G-77miOJ01WYD/view?usp=sharing">Demo Full Screen</a> & <a href="https://drive.google.com/file/d/1aDMZ_b7GVu_2cxiqC0h6sX1fOEz3iLD0/view?usp=sharing" target="__blank">[Demo Popup Window]</a>

</details>
New hooks added
- useCancelTransaction: a hook for both determining if a transaction can be canceled (account has enough funds to cover cancelation fee) and creating a memoized function for kicking off the cancelation workflow
- useRetryTransaction: a hook for determining the eligibility of retrying a transaction and creating a memoized function to kick off the retry workflow
- useI18nContext: a hook that is just a shortcut for importing both `useContext` and `I18nContext`
- useMethodData: a hook for tapping into token method data. Includes an effect to make sure that method data is requested if it is not already in `knownMethodData`
- useTokenData: a hook for memoized computation of transaction params data on token transfers
- useTransactionDisplayData: A hook for analyzing a transaction group and deriving a data shape that can be used for displaying a transaction to the user.

added utility methods
- formatDateWithYearContext: the new design calls for displaying the transaction year only if it did not occur this year. This method just provides a way for specifying two different format strings, one to use for current year transactions and a fallback to use for previous years transactions

refactored ListItem component
- Previously the ListItem component was very specific to transactions
- Moved all of the transaction-related logic into transaction-list-item
- Added new props to make the list item component more versatile
- Refactored component to be a stateless functional component

refactored TransactionList component
- Refactored to a functional component
- Combined container logic into component using hooks
- Updated the instantiation of TransactionListItems to match new component API

refactored TransactionListItem component
- Refactored to a functional component
- Combined and greatly modified container logic into the component using hooks
- Prepares all props needed to render a ListItem with transaction data

moved TransactionListItemDetails into a Popover
- Very minimal changes
- Dropped some media queries that adapted the detail view to larger display sizes as its always now in a Popover
- the new design of this view is not yet implemented, will do that in a new PR 
- Modified tests so that the existing tests are run against the first child of the primary wrapper (Runs tests against the original intended target instead of the new Popover)

### NEW UX FEATURES
<details>
<summary>Queued Transactions</summary>
When users have more than one pending transaction now all items beyond the first that aren't "Unapproved" transactions will have a "Queued" label and will not have the Loading indicator behind them. 

<img src="https://user-images.githubusercontent.com/4448075/81868902-b6743e00-9538-11ea-9c84-b63d6bf4aacf.png" alt="image" style="max-width:100%;">

</details>

<details>
<summary>Incremental loading of history</summary>
As pointed out in #8572 we are eagerly loading transactions into the history view, but we don't need to. We now have an incremental loader that shows 10 items at a time with a view more button to load more. 

<img src="https://user-images.githubusercontent.com/4448075/81868983-de63a180-9538-11ea-9003-914f8b360ddf.png" alt="image" style="max-width:100%;">

</details>

updated e2e test selectors for finding currency amounts in the list

### Bugs fixed
fixes #8572

whymarrh commented:
@brad-decker do you think it would be possible to address #8572 here?

brad-decker commented:
I’ll see what can be done @whymarrh I was thinking the same when I saw the issue report

brad-decker commented:
@whymarrh this now includes a resolution for #8572 - transaction-list.component now includes a limit for completed transactions and slices through the array of completed transactions. A new "View More" (pending @rachelcope 's feedback) button at the bottom of the transaction list loads an increment of transactions. I have created a constant PAGE_INCREMENT value of 10 for this purpose, but I expect that we could add a new user preference for this in the future (or perhaps 100 is a more reasonable default?)

brad-decker committed:
implement new transaction history design

whymarrh commented on ui/app/hooks/useTokenData.js:
```diff
@@ -0,0 +1,10 @@
+  if (!transactionData) {
+    return null
+  }
+  const tokenData = useMemo(() => getTokenData(transactionData), [transactionData])
+  return tokenData
```
Nit:

```suggestion
  return useMemo(() => getTokenData(transactionData), [transactionData])
```

whymarrh commented on ui/app/hooks/tests/useRetryTransaction.test.js:
```diff
@@ -0,0 +1,125 @@
+let useDispatch, useMethodData, useMetricEvent
+
+const dispatch = sinon.spy(() => Promise.resolve({ blockTime: 0 }))
+const trackEvent = sinon.spy()
```
Can we move these into the describe block scope instead of module scope and maybe also use a [sandbox](https://sinonjs.org/releases/v9.0.2/sandbox/) to simplify the restoration?

Gudahtt replied:
I definitely agree re: describe-scope, but I don't follow why using a sandbox would be simpler. Calling `sinon.restore()` in `afterEach` (thus resetting the "global sandbox") should be our default.

I've been following that pattern in an attempt to work toward being able to use a global `afterEach` hook for our entire test suite, [as recommended by the sinon docs](https://sinonjs.org/releases/v9.0.2/general-setup/), to completely eliminate the possibility of us forgetting to restore.

whymarrh replied:
Oh, that works too

whymarrh commented on ui/app/hooks/tests/useCancelTransaction.test.js:
```diff
@@ -0,0 +1,105 @@
+
+let useSelector, useDispatch
+
+const dispatch = sinon.spy()
```
Can we move this into the describe block scope instead of module scope and maybe also use a [sandbox](https://sinonjs.org/releases/v9.0.2/sandbox/) to simplify the restoration?

whymarrh commented on ui/app/components/app/transaction-list-item/index.scss:
```diff
@@ -1,149 +1,49 @@
-    &--show {
-      max-height: 1000px;
-      transition: max-height 700ms ease-out;
+  &__pendingActions {
```
Can we use kebab case for these selectors?

e.g.:

```suggestion
  &__pending-actions {
```

whymarrh replied:
Similarly for:

- `__primaryCurrency`
- `__secondaryCurrency`

brad-decker replied:
i also fixed a camelCase classname on the list component from the previous pr, but its basically unrelated to this work.

whymarrh reviewed (changes requested):
I only did a cursory read through but this looks good so far! I've left a few small comments.

brad-decker committed:
Update ui/app/hooks/useTokenData.js

Co-authored-by: Whymarrh Whitby <whymarrh.whitby@gmail.com>

brad-decker committed:
changes

brad-decker commented:
@whymarrh fixed those change reqs up :) thanks for the feedback.

Gudahtt commented on ui/app/hooks/useTransactionDisplayData.js:
```diff
@@ -0,0 +1,118 @@
+    subtitle += ` · ${initialTransaction.origin}`
+  } else if (transactionCategory === 'incoming') {
+    category = DEPOSIT
+    title = t('deposit')
```
Are we calling incoming transactions "Deposit"? I didn't see this in the design - it was called `Receive` there.

brad-decker replied:
I have no idea where I pulled this from, I'll update to Receive.

Gudahtt replied:
I'm pretty sure they're called "Deposit" on develop at the moment? That's probably where it came from.

brad-decker replied:
Updated to receive 👍

Gudahtt commented on ui/app/hooks/useTransactionDisplayData.js:
```diff
@@ -0,0 +1,118 @@
+  const { from: senderAddress, to } = initialTransaction.txParams || {}
+
+  // for smart contract interactions, methodData can be used to derive the name of the action being taken
+  const methodData = useSelector((state) => getKnownMethodData(state, primaryTransaction?.txParams?.data)) || {}
```
Previously this had used the initial transaction, not the primary transaction. `mergeProps` in `transaction-list-item.container.js` passes the initial transaction as `transaction`, which is passed through to `transaction-action.component.js`.
```suggestion
  const methodData = useSelector((state) => getKnownMethodData(state, initialTransaction?.txParams?.data)) || {}
```

Gudahtt replied:
..... 🤦‍♂️ Maybe not.

It looks like the old `TransactionAction` component had taken the `methodData` of the primary transaction, but had used the initial transaction when looking for the action key. Oops!

... I'm having trouble understanding why that would be desirable though.

Gudahtt replied:
Apparently this was broken in #6551 . It was supposed to use the initial transaction, as it did before that PR.

brad-decker replied:
![200w_d](https://user-images.githubusercontent.com/4448075/82695591-0040e080-9c2b-11ea-8ace-95671835506c.gif)

brad-decker replied:
so your suggestion is ... valid still?

Gudahtt replied:
Yeah I think so 😅

Gudahtt commented on ui/app/hooks/useTransactionDisplayData.js:
```diff
@@ -0,0 +1,118 @@
+  const tokenDisplayValue = useTokenDisplayValue(initialTransaction?.txParams?.data, token)
+
+  let category = SEND
+  let title = t('sendETH')
```
Rather than leaving this as the default, maybe we should ensure we exhaustively check all transaction categories? It seems dangerous to risk unidentified things being shown as a send.

brad-decker replied:
Is there an exhaustive list of transaction categories that could be seen here? I tried to piece them together through the various transforms that happen before we get to this point in the logic flow. I'm sure I'm missing more in retrospect. eg: what does a 'signing request' category look like?

Gudahtt replied:
I think this one function is the only place that `transactionCategory` is set for outgoing transactions: https://github.com/MetaMask/metamask-extension/blob/develop/app/scripts/controllers/transactions/index.js#L692
The the category is always `incoming` for incoming transactions (handled in the `incoming-transactions` controller).

Just 7 categories. I think those are all accounted for here in these 4 (as _send_ and _receive_ each include both ETH and tokens, and a contract deployment is displayed as a " contract interaction").

Then there are the non-transactions (which have no `transactionCategory`). I think this just includes signature requests, but I'm not 100% certain 🤔

Gudahtt replied:
It looks like signature requests are indeed the only non-transaction things included in the list. The transaction selector brings them in via `unapprovedMessagesSelector` - those are all signature requests.

brad-decker replied:
```js
  let category
  let title
  // There are four types of transaction entries that are currently differentiated in the design
  // 1. Send (sendEth sendTokens)
  // 2. Deposit
  // 3. Site interaction
  // 4. Approval
  if (transactionCategory == null) {
    category = APPROVAL
    title = t('signatureRequest')
    subtitle += ` · ${initialTransaction.origin}`
  } else if (transactionCategory === 'approve') {
    category = APPROVAL
    title = t('approveSpendLimit')
    subtitle += ` · ${initialTransaction.origin}`
  } else if (transactionCategory === 'contractDeployment' || transactionCategory === 'contractInteraction') {
    category = INTERACTION
    title = (methodData?.name && camelCaseToCapitalize(methodData.name)) || (actionKey && t(actionKey)) || ''
    subtitle += ` · ${initialTransaction.origin}`
  } else if (transactionCategory === 'incoming') {
    category = DEPOSIT
    title = t('deposit')
    prefix = ''
    subtitle += ` · From: ${shortenAddress(senderAddress)}`
  } else if (transactionCategory === 'transfer') {
    category = SEND
    title = t('sendSpecifiedTokens', [token?.symbol])
    recipientAddress = getTokenToAddress(tokenData.params)
    subtitle += ` · To: ${shortenAddress(recipientAddress)}`
  } else if (category === SEND) {
    category = SEND
    title = t('sendETH')
    subtitle += ` · To: ${shortenAddress(recipientAddress)}`
  }
```

this is the adjustment I've made. Not sure if signing requests should be approvals or contract interactions for now, or hold off for new direction from @rachelcope on a new symbol to display. 

two questions @Gudahtt 
1: do you think, pending design feedback, this covers our bases.
2: This isn't much better than setting category, subtitle and title to the 'signature request' as defaults... should we find a way to update transactionCategory for signature requests throughout the system so that its not a fallback?

Gudahtt replied:
Until we get a better icon to use, I'd suggest using the icon we use today for signature requests for now (the Ethereum logo). As inappropriate and non-explanatory as the Ethereum logo is for these, each of these new four would be equally inappropriate, so :man_shrugging: 

Assuming anything without a category is a signature request seems fine to me for now. Certainly assigning them a transaction category seems inappropriate. A better solution might be creating a separate `ListItem` for signature requests, and splitting up the selectors and everything to treat them as separate. But that's a ton of work that we can think about some other day.

RE: Having an exhaustive list of transactions, I mainly wanted to make sure we weren't missing any categories.

brad-decker replied:
![image](https://user-images.githubusercontent.com/4448075/82685837-77ba4400-9c1a-11ea-8aca-14d433f0bde2.png)

brad-decker committed:
Update ui/app/hooks/useTransactionDisplayData.js

Co-authored-by: Mark Stacey <markjstacey@gmail.com>

brad-decker committed:
address change requests

brad-decker commented on ui/app/components/app/transaction-list-item/transaction-list-item.component.js:
```diff
@@ -1,278 +1,172 @@
+          senderAddress={senderAddress}
+          recipientAddress={recipientAddress}
+          onRetry={retryTransaction}
+          showRetry={isEarliestNonce && isPending}
```
updated!

brad-decker commented on ui/app/hooks/useTransactionDisplayData.js:
```diff
@@ -0,0 +1,118 @@
+  let prefix = '-'
+  const date = formatDateWithYearContext(initialTransaction.time || 0)
+  let subtitle = date
+  let recipientAddress = to
```
@Gudahtt are we good with the way it is written now? I agree we shouldn't assume the position of params unless we've validated we are working with a token transaction first.

Gudahtt replied:
That depends on whether this constitutes a functional change or not, and if so what that change is. I'm not sure yet. I'll take another look.

Gudahtt replied:
Hmm, yeah the old way of doing this was downright irresponsible. This'd show the wrong recipient for_most_ ERC20 methods 😨. It was only correct for `transfer`, but it didn't verify it was a transfer.

Gudahtt replied:
The way you wrote it is definitely better, in that it's not totally broken. I guess I'll mark this as resolved then.

brad-decker commented:
@Gudahtt -- So I am digging into failing integration tests. With the new logic in place the integration tests render all transactions as 'Signature Request' because none of the transactions in the sample data have transactionCategory set.    So I think the right approach here is to go through the data and add the appropriate categories (I'm a little daunted by this) and then update the test expectations. Do you agree or have a different path forward?

Gudahtt commented:
@brad-decker That approach makes sense to me (unfortunately 😅 ) They should all have a category  - all real transactions should have one, as far as I know anyway.

Gudahtt commented on ui/app/hooks/useTransactionDisplayData.js:
```diff
@@ -0,0 +1,124 @@
+    category = SIGNATURE_REQUEST
+    title = t('signatureRequest')
+    subtitle += ` · ${initialTransaction.msgParams?.origin || initialTransaction.origin || ''}`
+  } else if (transactionCategory === 'approve') {
```
All of these action categories are declared as constants in `ui/app/helpers/constants/transactions.js`. Better to use the constants here instead of the string literals.

brad-decker committed:
add transferfrom, care for lack of origin on sig req

brad-decker committed:
add categories and tokens to integration tests

Gudahtt commented on ui/app/components/app/transaction-list-item/transaction-list-item.component.js:
```diff
@@ -1,278 +1,178 @@
+
+  const { title, subtitle, category, primaryCurrency, recipientAddress, secondaryCurrency, status, senderAddress } = useTransactionDisplayData(transactionGroup)
+
+  const isApprove = category === 'approval'
```
Same here re: using constants

brad-decker commented:
@Gudahtt / @MetaMask/extension-devs  -- fixed the integration tests by adding transaction categories. To make the tests reflect the existence of token symbols I added a couple of fake tokens to the tokens state key in the mocked data. Then I updated the tests to look for the appropriate things.

Gudahtt commented on ui/app/hooks/useTransactionDisplayData.js:
```diff
@@ -0,0 +1,125 @@
+  const primaryValue = primaryTransaction.txParams?.value
+  let prefix = '-'
+  const date = formatDateWithYearContext(initialTransaction.time || 0)
+  let subtitle = date
```
The `subtitle` in the designs shows as `Failed` for failed transactions, and as either `Pending` or `Queued` for pending transactions

brad-decker replied:
Cool. the way i had it the subtitles were a little lengthy!

Gudahtt commented on ui/app/hooks/useTransactionDisplayData.js:
```diff
@@ -0,0 +1,125 @@
+    category = RECEIVE
+    title = t('receive')
+    prefix = ''
+    subtitle += ` · From: ${shortenAddress(senderAddress)}`
```
These `From` and `To` string literals should be using localized strings, with the address as input.

The ` · ` part seems fine as a string-literal though.

brad-decker committed:
use constants and I18n on strings

brad-decker commented:
@Gudahtt your latest feedback addressed in latest commit.

Gudahtt commented on ui/app/hooks/useTransactionDisplayData.js:
```diff
@@ -0,0 +1,152 @@
+  const [primaryCurrency] = useCurrencyDisplay(primaryValue, {
+    prefix,
+    displayValue: tokenDisplayValue,
+    suffix: token?.symbol,
```
Similar to the above comment; we should make sure not to include this for non-transfer transactions

Gudahtt replied:
Though the `secondaryCurrency` might be the one requiring a suffix depending on the user's preference, so the `suffix` might be required on the `secondaryCurrency` instead of the `primaryCurrency` 🤔

brad-decker replied:
The secondaryCurrency will never display a token value unless I'm mistaken...

Gudahtt replied:
Right, but that's the problem. Or at least it seems like a problem - maybe it's intentional 🤷‍♂️.

`useUserPreferencedCurrency` determines whether fiat is "primary" or "secondary", and from there the `displayValue` overrides that to show the token value. So something is wonky here; right now the primary will always show the token, and the secondary will show either fiat or eth depending on user's preferred currency. But it'll show whichever one _isn't_ their preferred currency. So a user that prefers ETH will be shown fiat as secondary, and vice versa.

I'm not sure what we want to do here 🤔 

Either:

1) Show the token value in whichever place ETH would normally be shown
2) Show the token value first, with the user's preferred currency second
3) Show the token value first, with the user's non-preferred currency second
4) Show the token value alone.

brad-decker replied:
I may be confused -- is this something that differs from the current develop branch or are we discussing an improvement that shows relative eth value or fiat value for the token?

Gudahtt replied:
The current develop branch doesn't show the token value at all here. I thought you had implemented #3 at the time I wrote that comment, which seems kinda wrong. But since your latest update, it's #4, which seems just fine for the time being :+1:

Gudahtt replied:
Er... I guess the original comment here is still not resolved, though the tangential topic I brought up re: the secondary currency is.

I think this should still depend upon `isTokenCategory` being true:

```suggestion
    suffix: isTokenCategory && token?.symbol,
```

brad-decker replied:
Could we jump on a call today at some point to walk through this? I’m still not following what is unintended and what the issue is with this logic

Gudahtt replied:
Sure! Though, in a screenshot:

![dia-with-fiat](https://user-images.githubusercontent.com/2459287/82904675-2ae7ad80-9f39-11ea-8c4e-2a97586b9708.PNG)

The current designs show a fiat secondary balance. It's unclear to me whether this was supposed to _always_ show fiat, but it's certainly different from this PR which omits the secondary balance altogether.

brad-decker replied:
Right now I'm going to leave it as always and only showing the token value. I'll follow up with @rachelcope about whether it was intended to show fiat currency for token transactions, but if we do that it can't use our existing logic and would be a hefty lift because we don't have (to my knowledge) exchange rates for tokens yet? If we showed Fiat here would it be the current going rate for those tokens or would it be the rate at the time of purchase?

Gudahtt replied:
We do have exchange rates for tokens in the send flow, and probably other places as well. It was never included in the transaction list though.

I agree that omitting this for now makes sense :+1:

brad-decker commented on ui/app/hooks/useTransactionDisplayData.js:
```diff
@@ -0,0 +1,152 @@
+
+  const token = knownTokens.find((token) => token.address === recipientAddress)
+  const tokenData = useTokenData(initialTransaction?.txParams?.data)
+  const tokenDisplayValue = useTokenDisplayValue(initialTransaction?.txParams?.data, token)
```
The problem is you can't use hooks inside logic blocks, they have to be top level.

brad-decker replied:
That is why the useTokenData and useTokenDisplayValue hooks care for the cases in which they are not needed by returning null at the top.

Gudahtt replied:
Ah, right. Using them could certainly be contingent upon the transaction being a token transfer though.

brad-decker replied:
`tokenData` is only used in the case of the transaction being a `transfer` or `transferfrom` category. `tokenDisplayValue` is passed to the `useCurrencyDisplay` hooks which also have to be top-level, but if `tokenDisplayValue` is undefined/null it won't be used by that hook.

Gudahtt replied:
`tokenDisplayValue` might be truthy for transactions that aren't token transfers though. In those cases, the `primaryCurrency` would be displayed incorrectly.

Maybe I made this comment on the wrong line - this might be better solved by adding some conditions to whether these are passed to `useCurrencyDisplay`

brad-decker replied:
I think this is now resolved by explicitly passing undefined when not dealing with a token transfer transactionCategory

brad-decker commented on ui/app/hooks/useTransactionDisplayData.js:
```diff
@@ -0,0 +1,125 @@
+    subtitle += ` · From: ${shortenAddress(senderAddress)}`
+  } else if (transactionCategory === 'transfer' || transactionCategory === 'transferfrom') {
+    category = SEND
+    title = t('sendSpecifiedTokens', [token?.symbol])
```
Good call.

brad-decker replied:
resolved.

brad-decker committed:
show the last page of results

brad-decker committed:
make token related logic clearer

brad-decker commented on ui/app/hooks/useTransactionDisplayData.js:
```diff
@@ -111,7 +117,7 @@ export function useTransactionDisplayData (transactionGroup) {
     prefix = ''
     subtitle = `${t('from')}: ${shortenAddress(senderAddress)}`
-  } else if (transactionCategory === TOKEN_METHOD_TRANSFER || transactionCategory === TOKEN_METHOD_TRANSFER_FROM) {
+  } else if () {
```
oops

brad-decker committed:
reword comment

brad-decker committed:
add fallback for missing token data

brad-decker committed:
use correct category for approve

brad-decker committed:
approval uses tokenData

Gudahtt commented on ui/app/helpers/constants/transactions.js:
```diff
@@ -23,3 +24,12 @@ export const ENCRYPTION_PUBLIC_KEY_REQUEST_KEY = 'encryptionPublicKeyRequest'
 export const CONTRACT_INTERACTION_KEY = 'contractInteraction'
 export const CANCEL_ATTEMPT_ACTION_KEY = 'cancelAttempt'
 export const DEPOSIT_TRANSACTION_KEY = 'deposit'
+export const INCOMING_TRANSACTION_KEY = 'incoming'
```
Nit: Calling this a "key" is a bit misleading, because this isn't the key for any localized messages.

It's quite unfortunate how these constants were originally written - first as keys to localized messages about specific transaction categories, then later re-used as transaction category was added as a real `txMeta` property. Also the pattern of using variables as the parameter to the translation method makes static analysis of localized messages much more difficult. Ah well.

For now I'd suggest removing `_KEY` from this variable name, and leaving the rest as-is. But at some point I'd like to return here and stop using these as translation keys altogether, instead using these just as transaction category enums.

brad-decker replied:
I wasn’t even quite sure what _KEY was referring to here and didn’t pick up the significance of it mapping to a localization key. I was attempting to keep the patterns of the file in place but I don’t have any objections to your request.

brad-decker replied:
I believe this is resolved now.

brad-decker commented on ui/app/hooks/useTransactionDisplayData.js:
```diff
@@ -0,0 +1,169 @@
+}
+
+const tokenCategoryMap = {
+  [TOKEN_METHOD_APPROVE]: true,
```
Yes, it is used to show the spend limit which it has done historically as well. The original submission prior to changes from feedback worked to display the spend limit, but once I added the explicit bypassing of token related variables e2e tests started failing. What had happened was exactly what you were hoping to prevent or surface : the token related variables were being used in places I hadn’t expected and when I became more explicit about passing them in the approve workflow broke. So this was added to be able to correctly use the token data/displayValue for approval

Gudahtt replied:
Ah I see. I was assuming that we _didn't_ want to show the spend limit in the transaction list, as the new design kinda implies only the ETH cost is shown? But it is ambiguous, and this is how it works already, so this seems fine for now. We can ask for clarity on this later. I think you asked a similar question in a Figma comment.

brad-decker commented on ui/app/hooks/useRetryTransaction.js:
```diff
@@ -0,0 +1,110 @@
+
+
+  const showRetryTransactionDialog = useCallback((event) => {
+    event.preventDefault()
```
No, this should be safe to remove. Stop propagation is required to prevent opening the transaction details when interacting with buttons

brad-decker replied:
removed! :)

brad-decker commented on ui/app/helpers/constants/transactions.js:
```diff
@@ -7,14 +7,15 @@ export const CONFIRMED_STATUS = 'confirmed'
 
 export const SEND_ETHER_ACTION_KEY = 'sentEther'
 export const DEPLOY_CONTRACT_ACTION_KEY = 'contractDeployment'
-export const APPROVE_ACTION_KEY = 'approve'
```
@Gudahtt did this before I got more comfortable with the internationalization system -- though I'm still torn here. The text for the approve message is changing from 'Approve' to 'Approve Spending Limit' -- it seems like changing the text would be quite the break from the current translations but it would still capture the essence of the message ... should I just rephrase the en/messages.json "message" for the 'approve' key instead of adding a new key?

Gudahtt replied:
Hmm. Yeah that seems fine to me 🤔 

I wish we had a better way of handling changes to these messages. Leaving the "Approve" transactions does seem better than erasing them though.

brad-decker replied:
approve action key restored to 'approve'
https://github.com/MetaMask/metamask-extension/blob/b695366e0c978cd969e86ba38eb8a7396b2c8daf/ui/app/helpers/constants/transactions.js#L36

brad-decker commented on ui/app/hooks/useRetryTransaction.js:
```diff
@@ -0,0 +1,110 @@
+  }))
+  const dispatch = useDispatch()
+  const { fetchBasicGasAndTimeEstimates, fetchGasEstimates, setSelectedToken, retryTransaction } = useMemo(() => ({
+    fetchBasicGasAndTimeEstimates: () => dispatch(fetchBasicGasAndTimeEstimatesAction()),
```
It prevents the other three functions from being created every render, only recreating them when the retryTransaction method must also be constructed. This is close to the behavior of mapDispatchToProps. ``

Gudahtt replied:
Technically it does re-create the functions every render though - it just doesn't use them. The only performance advantage here is in preserving the referential equality of this function in child components (i.e. preventing re-renders), but I don't think we _use_ these callbacks in that matter, so... this is more work for no gain, I think? Unless I'm misunderstanding something here.

Here's the explanation I was thinking of: https://kentcdodds.com/blog/usememo-and-usecallback

Gudahtt replied:
Separately from that, this caught my eye originally because of how these functions were grouped together. It would seem better to have them in separate `useCallback` hooks, so that each is only re-created when necessary instead of them _all_ being recreated when any dep changes.

brad-decker replied:
@Gudahtt 🤦‍♂️ just when you think you know something. Thanks for sharing the article. The truth is wrapping those three other functions is just mimicking the mapDispatchToProps... there's almost nothing to gain from actually creating a new function that just calls dispatch because I'm not passing the functions anywhere.. i could just call dispatch directly where I am calling these methods in the code... which means even less memory allocation because the selectors are instantiated once, dispatch is allocated by useDispatch...

brad-decker replied:
Lemme know what you think of the new flow @Gudahtt -- just using dispatch directly instead of creating the intermediary functions

Gudahtt replied:
Works for me!

brad-decker commented on ui/app/hooks/useRetryTransaction.js:
```diff
@@ -0,0 +1,110 @@
+    return Date.now() - submittedTime > 5000 && isEarliestNonce && !hasRetried
+  })
+
+  if (methodData?.name === TOKEN_METHOD_TRANSFER && initialTransaction.txParams?.to) {
```
I think you're right it should be moved into the callback. Though I'm not entirely sure what the original purpose of this call was here: https://github.com/MetaMask/metamask-extension/blob/fd170ef8c0404b5260a36dba38245e629949c6d5/ui/app/components/app/transaction-list-item/transaction-list-item.component.js#L111-L125

brad-decker replied:
why would you set a selected token as a result of issuing a retry?

Gudahtt replied:
That is an excellent question 😅 😬 

I'm not sure. This is one of the things I had planned to remove actually - in my asset page PR (I need to replace all uses of `selectedToken` with a token prop that gets passed in).

It's possible we're using the selected token somewhere in the retry logic? That doesn't make sense though 🤔. I'll look into this... maybe this can be omitted now...

brad-decker replied:
@Gudahtt i didn't touch this on this round of fixes.. lemme know if we're good to either remove or move it into callback

brad-decker commented on ui/app/hooks/useRetryTransaction.js:
```diff
@@ -0,0 +1,110 @@
+      if (Date.now() - submittedTime > 5000) {
+        setRetryEnabled(true)
+      } else {
+        intervalId = setInterval(() => {
```
I like measuring and delaying an exact amount way better. Not sure why I went immediately for an interval

brad-decker replied:
https://github.com/MetaMask/metamask-extension/blob/b695366e0c978cd969e86ba38eb8a7396b2c8daf/ui/app/hooks/useShouldShowSpeedUp.js#L25-L33

whymarrh commented on ui/app/hooks/useMethodData.js:
```diff
@@ -0,0 +1,30 @@
+import { getKnownMethodData } from '../selectors/selectors'
+
+/**
+ * useMethodData
```
Nit: can we add a single line description of the function here instead? Maybe extracted from the larger description below.

brad-decker replied:
Do you like: 
``` * Access known method data and attempt to resolve unknown method data ``` for this?

whymarrh commented on ui/app/hooks/useRetryTransaction.js:
```diff
@@ -0,0 +1,110 @@
+
+
+/**
+ * useRetryTransaction
```
Nit: can we add a single line description of the function here instead? Maybe extracted from the larger description below.

brad-decker replied:
the description is now short enough to meet this need. Should I just omit the name? new:

```
/**
 * Provides a reusable hook that, given a transactionGroup, will return
 * a method for beginning the retry process
 * @param {Object} transactionGroup - the transaction group
 * @return {Function}
 */
```

whymarrh commented on ui/app/hooks/useTransactionDisplayData.js:
```diff
@@ -0,0 +1,169 @@
+ */
+
+/**
+ * useTransactionDisplayData
```
Nit: can we add a single line description of the function here instead? Maybe extracted from the larger description below.

brad-decker replied:
```
/**
 * Get computed values used for displaying transaction data to a user
 *
```

whymarrh commented on ui/app/hooks/useCancelTransaction.js:
```diff
@@ -0,0 +1,50 @@
+import { getConversionRate, getSelectedAccount } from '../selectors'
+
+/**
+ * useCancelTransaction
```
Nit: can we add a single line description of the function here instead? Maybe extracted from the larger description below.

brad-decker replied:
```
/**
 * Determine whether a transaction can be cancelled and provide a method to
 * kick off the process of cancellation.
```

brad-decker committed:
resolving feedback items from 5/26

brad-decker commented on ui/app/hooks/useTransactionDisplayData.js:
```diff
@@ -0,0 +1,169 @@
+    category = TRANSACTION_CATEGORY_RECEIVE
+    title = t('receive')
+    prefix = ''
+    subtitle = `${t('from')}: ${shortenAddress(senderAddress)}`
```
https://github.com/MetaMask/metamask-extension/blob/b695366e0c978cd969e86ba38eb8a7396b2c8daf/app/_locales/en/messages.json#L1524-L1527

https://github.com/MetaMask/metamask-extension/blob/b695366e0c978cd969e86ba38eb8a7396b2c8daf/app/_locales/en/messages.json#L666-L669

brad-decker replied:
and the usage:
https://github.com/MetaMask/metamask-extension/blob/b695366e0c978cd969e86ba38eb8a7396b2c8daf/ui/app/hooks/useTransactionDisplayData.js#L109-L123

Gudahtt replied:
Looks great!

brad-decker commented on ui/app/hooks/useTransactionDisplayData.js:
```diff
@@ -0,0 +1,169 @@
+// from the history view we will already know the pending status
+// of a transaction, but in order to make this hook less coupled
+// to where it is used we should determine the pending status here
+const pendingStatusMap = {
```
https://github.com/MetaMask/metamask-extension/blob/b695366e0c978cd969e86ba38eb8a7396b2c8daf/ui/app/helpers/constants/transactions.js#L11-L32

removed the pending status that i added per your suggestion, and moved the hash maps into the constants file.

brad-decker commented on ui/app/hooks/useRetryTransaction.js:
```diff
@@ -0,0 +1,110 @@
+ * @param {boolean} isEarliestNonce - indicates if this transaction is the earliest in history
+ * @return {[boolean, Function]}
+ */
+export function useRetryTransaction (transactionGroup, isEarliestNonce = false) {
```
https://github.com/MetaMask/metamask-extension/blob/transaction-detail-and-history/ui/app/hooks/useRetryTransaction.js probably easier to just link the file, removed all the speed up (i think) logic here to keep it purely associated with retry... there's a new hook now:

https://github.com/MetaMask/metamask-extension/blob/transaction-detail-and-history/ui/app/hooks/useShouldShowSpeedUp.js

of particular interest:

https://github.com/MetaMask/metamask-extension/blob/b695366e0c978cd969e86ba38eb8a7396b2c8daf/ui/app/hooks/useShouldShowSpeedUp.js#L25-L33

brad-decker commented on ui/app/helpers/constants/transactions.js:
```diff
@@ -7,14 +7,15 @@ export const CONFIRMED_STATUS = 'confirmed'
 export const FAILED_STATUS = 'failed'
 export const DROPPED_STATUS = 'dropped'
 export const CANCELLED_STATUS = 'cancelled'
+export const PENDING_STATUS = 'pending'
```
Removed in favor of the hash table that is used in useTransactionDisplayData and the nonceSorted transaction list selectors

brad-decker commented on ui/app/hooks/useRetryTransaction.js:
```diff
@@ -0,0 +1,110 @@
+    retryTransaction: () => {
+      const transaction = initialTransaction
+      const increasedGasPrice = increaseLastGasPrice(gasPrice)
+      dispatch(setCustomGasPriceForRetryAction(increasedGasPrice || transaction.txParams?.gasPrice))
```
in this specific case or anywhere? e2e failed pretty heavily with errors about accessing a value on undefined with txParams. I'll remove this and see if i can get you better details.

Gudahtt replied:
I _think_ `txParams` should always exist on actual transactions? It'd be missing from signature requests, but I don't see how this hook would get used for those.

brad-decker replied:
@Gudahtt its probably signing requests because hooks are all called at top level and some work is done inside the top level of this hook:

```
export function useRetryTransaction (transactionGroup) {
  const { primaryTransaction, initialTransaction } = transactionGroup
  const gasPrice = primaryTransaction.txParams?.gasPrice
```
i probably then got overzealous by adding ? marks anywhere I saw txParams. it should always be defined inside this callback. I'll see if i can remove this one.

Gudahtt replied:
Ah right. Ok, that would make sense. So this would still get run for signature request items, but the callback would never be used.

Keeping the `?.` is totally fine with me - I was just curious.

brad-decker commented on ui/app/hooks/useRetryTransaction.js:
```diff
@@ -0,0 +1,110 @@
+    }
+
+    trackMetricsEvent()
+    return fetchBasicGasAndTimeEstimates()
```
Implemented async/await here

brad-decker commented on ui/app/hooks/useRetryTransaction.js:
```diff
@@ -0,0 +1,110 @@
+
+    if (!retryEnabled) {
+      // Eject here because the function was called erroneously
+      return
```
made it async, also retryEnabled went away because of the removing of the speed up logic.

brad-decker committed:
restore original logic of setting selected token

brad-decker commented:
@whymarrh and @Gudahtt --  that last change was made and tests passing. 🤞

Gudahtt commented on ui/app/hooks/useCancelTransaction.js:
```diff
@@ -0,0 +1,50 @@
+    conversionRate,
+  })
+
+  const cancelTransaction = useCallback((event) => {
```
^ Not a blocker, but I do think this could be removed

Gudahtt reviewed (approved):
LGTM!

whymarrh reviewed (approved):
:shipit:

Thanks for championing this across the finish line! This is a huge improvement. 🎉
